spec/src/main/asciidoc/servlet-spec-body.adoc

```diff
@@ -1303,6 +1303,145 @@ the header value to an `int`, a `NumberFormatException` is thrown. If
+
+. **Concatenate segments.**
++
+The segments are concatenated into a single path string with each segment preceded by the `"/"` character. If the original path after stripping query and fragment ends with a `"/"` then append a final `"/"` character.
+ If a segment contains the `"/"` or `"%"` characters, and the request will not be rejected below, then the container may configured re-encode those characters to the `%nn` form. If any characters are re-encoded, then the `"%"` must also be re-encoded.
+
+. **Mapping URI to context and resource.**
++
+The decoded path is used to map the request to a context and resource within the context. This form of the URI path is used for all subsequent mapping (web applications, servlet, filters and security constraints).
+
+. **Rejecting Suspicious Sequences.**
++
+If suspicious sequences are discovered during the prior processing steps, the request must be rejected with a 400 bad request rather than dispatched to the target servlet. If a context is matched then the error handling of the context may be used to generate the 400 response. By default, the set of suspicious sequences below, but may be configured differently by a container:
+
+  * Any path not starting with the `"/"` character
+  * Any path starting with an initial segment of `".."`
+  * The encoded `"/"` character
+  * Any `"."` or `".."` segment that had a path parameter
+  * Any `"."` or `".."` segment with any encoded characters
+  * Any `".."` segment preceded by an empty segment
+  * The `"\"` character encoded or not.
+  * Any control characters either encoded or not.
+
+==== Example URIs
+> TODO more work needed on this section. Perhaps some parts need to be moved to other sections.
+
+Consider the following URIs with a forbidden security constraint at `/secret/*` and a servlet registered at /dispatch/* that forwards request to the string return from pathInfo()
+
+. Example URIs
+|===
+| Encoded URI path  | Decoded Path | Rejected
+
+| `/public/.;/file.txt` | `/public/file.txt/` | 400 dot segment with parameter
+| `/public/file.txt` | `/public/file.txt/` |
+| `/public//file.txt` | `/public/file.txt/` |
+| `/public/;/file.txt` | `/public/file.txt/` |
+| `/PUBLIC/file.txt` | `/PUBLIC/file.txt/` |
+| `/public%2Ffile.txt` | `/public/file.txt/` | 400 encoded /
+| `/public%5Cfile.txt` | `/public\file.txt/` | 400 backslash character
+| `/public%00/file.txt` | `/public /file.txt/` | 400 control character
+| `/public/./file.txt` | `/public/file.txt/` |
+| `/public/%2e/file.txt` | `/public/file.txt/` | 400 encoded dot segment
+| `/public/%2e;/file.txt` | `/public/file.txt/` | 400 encoded dot segment
+| `/../docroot/public/file.txt` | `/../docroot/public/file.txt/` | 400 leading dot-dot-segment
+| `/public/dir/../file.txt` | `/public/file.txt/` |
+| `/public//../file.txt` | `/file.txt/` |
+| `/public/;param/../file.txt` | `/file.txt/` |
```

Comment:
Looking at this example made me uneasy. Can we add `"/" sub-delim *(pchar)` to the list of suspicious sequences to be rejected?

Reply:
I already have 

> Any `".."` segment preceded by an empty segment

But that doesn't fire because we are removing empty segments before we normalize.   As said above, I'd prefer to remove empty segments last so this one would not be an issue.
But for now I'll add 

> Any empty segment with a parameter

Whilst we discuss the ordering.   I don't mind if ultimately we leave in suspect patterns that cannot happen by default, as they may still trigger if there are container specific configurations.  Thus containers will need to both configure normalization and explicitly allow suspect patterns for them to arrive at the application.

Reply:
This is now:
> | `/public/;param/../file.txt` | `/file.txt` | 400 empty segment with parameters

But I would prefer
> | `/public/;param/../file.txt` | `/public/file.txt` | 400 empty segment before dot dot

Reply:
Actually, I have convinced myself otherwise.  The reason we remove empty segments is that many file systems treat them the same as a '.' segment, therefore we should do the same.
So `/foo//bar` should be handled the same as `/foo/./bar` and be normalized to `/foo/bar`.    But normalization of `/foo/./../bar` by the RFC is (surprisingly) `/bar`, so `/foo//..bar` should normalize to the same result, as should `/foo/;/../bar`.